core/mempool/src/lib.rs

```diff
@@ -83,14 +87,22 @@ impl Pool {
     }
 
     pub fn add_transaction(&self, transaction: SignedTransaction) -> Result<(), String> {
-        if let Ok(Some(_)) = self
-            .storage
-            .write()
-            .expect(POISONED_LOCK_ERR)
-            .transaction_address(&transaction.get_hash())
         {
-            return Ok(());
+            // validate transaction
+            let mut guard = self.storage.write().expect(POISONED_LOCK_ERR);
+            let last_block_index = guard.blockchain_storage_mut().best_block().unwrap().unwrap().index();
+            if last_block_index > 0 {
+                let old_tx_nonce = *guard.tx_nonce(last_block_index).unwrap().unwrap();
+                if transaction.body.get_nonce() <= old_tx_nonce {
+                    return Ok(());
+                }
+            }
+            if let Ok(Some(_)) = guard.transaction_address(&transaction.get_hash())
```

Comment:
This seems like a redundant check now. `transaction_address` is only extended when the block is imported or produced. So `transaction` fails this check it wouldn't have passed the nonce check above.

Reply:
Good point. Removed